### PR TITLE
Fix SVM Assert caused by missing class validation

### PR DESCRIPTION
**Pull request by dsouzai**
The code in J9Inliner.cpp that implemented certain heuristics to choose a class to inline based on profiling missed adding some SVM validations. This PR fixes this by wrapping the code that finds a candidate class in a heuristic region, and then adding the necessary profiled class and instanceof validations.

These heuristics are also disabled for non-SVM AOT. Normally this code is implicitly disabled as it only runs post-startup; however, it is possible for the VM to transition from startup to notstartup in the middle of a compilation, which could result in this code running in a non-SVM compilation which is not capable of generating the necessary validations.

Fixes https://github.com/eclipse-openj9/openj9/issues/17736

### CODE REVIEW

**Comment by dsouzai** [comment]
@jdmpapin for your review.

File: runtime/compiler/optimizer/J9Inliner.cpp [hunk 1]
```diff
@@ -855,7 +855,7 @@ bool TR_J9InterfaceCallSite::findCallSiteTargetImpl(
          kind = TR_InterfaceGuard;
          testType = TR_MethodTest;
          }
-      else
+      else if (!comp()->compileRelocatableCode() || comp()->getOption(TR_UseSymbolValidationManager))
```
Comment (jdmpapin) [inline runtime/compiler/optimizer/J9Inliner.cpp:858]: I don't think we need to skip this entire section for non-SVM AOT. It should suffice to set `useVftTestHeuristics` to false

File: runtime/compiler/optimizer/J9Inliner.cpp [hunk 2]
```diff
@@ -948,16 +950,36 @@ bool TR_J9InterfaceCallSite::findCallSiteTargetImpl(
...
+                          && comp()
+                             ->getSymbolValidationManager()
+                             ->addClassInstanceOfClassRecord(topProfiledClass, iface, true, true, true)))
+                     {
+                     testType = TR_VftTest;
+                     thisClass = topProfiledClass;
+                     }
+                  else
+                     {
+                     calleeResolvedMethod = NULL;
```
Comment (jdmpapin) [inline runtime/compiler/optimizer/J9Inliner.cpp:973]: I don't think failure to create the records should require us to discard this call target. We can still inline it with a method test like we would have if we hadn't looked at the profiling

File: runtime/compiler/optimizer/J9Inliner.cpp [hunk 3]
```diff
@@ -948,16 +950,36 @@ bool TR_J9InterfaceCallSite::findCallSiteTargetImpl(
...
+               comp()->exitHeuristicRegion();
+
                if (topProfiledClass != NULL
                    && remainingTotalFreq >= 32
                    && topProfiledFreq == remainingTotalFreq)
                   {
-                  testType = TR_VftTest;
-                  thisClass = topProfiledClass;
+                  if (!comp()->compileRelocatableCode()
+                      || (comp()->getOption(TR_UseSymbolValidationManager)
```
Comment (jdmpapin) [inline runtime/compiler/optimizer/J9Inliner.cpp:960]: Would it be good to assert that SVM is in use if we reach this point in AOT?

**Comment by jdmpapin** [comment]
None of my comments just above are so serious that they should necessarily block merging - just let me know what you think

**Comment by dsouzai** [comment]
@jdmpapin I addressed the review comments since they were relatively straightforward to do and to not make non-SVM AOT unnecessarily conservative.

File: runtime/compiler/optimizer/J9Inliner.cpp [hunk 4]
```diff
@@ -924,36 +928,57 @@ bool TR_J9InterfaceCallSite::findCallSiteTargetImpl(
...
 
                if (topProfiledClass != NULL
                    && remainingTotalFreq >= 32
                    && topProfiledFreq == remainingTotalFreq)
                   {
-                  testType = TR_VftTest;
-                  thisClass = topProfiledClass;
+                  bool valid = true;
+
+                  if (!comp()->compileRelocatableCode())
```
Comment (jdmpapin) [inline runtime/compiler/optimizer/J9Inliner.cpp:970]: This condition is backwards

Comment (dsouzai) [reply runtime/compiler/optimizer/J9Inliner.cpp]: Fixed in force push.

**Review by jdmpapin — APPROVED** [review]

**Comment by jdmpapin** [comment]
Jenkins test sanity.functional+aot all jdk17

**Comment by jdmpapin** [comment]
The z test failure is #17228, and the AIX build failure is #8625. Ignoring z and restarting AIX

Jenkins test sanity.functional+aot aix jdk17

**Pull request merged**